docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -31,90 +31,116 @@ Here Semgrep tracks the data returned by `get_user_input()`, which is the source
+<iframe src="https://semgrep.dev/embed/editor?snippet=eqYN8" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-```javascript
+This is the default for historical reasons, but it may change in the future.
+
+It is possible to instruct Semgrep to only consider as taint sources the "exact" matches of a souce pattern by setting `exact: true`:
+
+```yaml
 pattern-sources:
-  - patterns:
-    - pattern-inside: function ($REQ, ...) {...}
-    - focus-metavariable: $REQ
+- pattern: source(...)
+  exact: true
 ```
 
+Once the source is "exact", Semgrep will no longer consider subexpressions as taint sources, and `sink(x)` inside `source(sink(x))` will not be reported as a tainted sink (unless `x` is tainted in some other way).
+
+<iframe src="https://semgrep.dev/embed/editor?snippet=Zq5ow" border="0" frameBorder="0" width="100%" height="432"></iframe>
+
+For many rules this distinction is not very meaningful because it does not always make sense that a sink occurs inside the arguments of a source function.
+
+:::note
+If one of your rules relies on non-exact matching of sources, we advice you to make it explicit with `exact: false`, even if it is the current default, so that your rule does not break if the default changes.
+:::
+
 ### Sources by side-effect
 
-Consider the following hypothetical Python code where we call a `make_tainted` function that makes its argument tainted by side-effect:
+Consider the following hypothetical Python code, where `make_tainted` is a function that makes its argument tainted by side-effect:
 
 ```python
-def make_tainted(a_set):
-  a_set.add(tainted)
-
 make_tainted(my_set)
 sink(my_set)
 ```
 
-Using the source specification below, the pattern formula only matches the occurrence of `x` that is the actual parameter of `make_tainted`, but it does not affect the occurrence of `x` that is passed to `sink`. Both are the same variable but in different occurrences. Only the first one is matched by the source specification. At the same time, Semgrep does not know that `make_tainted` is updating the variable `x` by side-effect. Thus, a taint rule using such specifications does not produce any finding.
+This kind of source can be specified by setting `by-side-effect: true`:
 
 ```yaml
 pattern-sources:
-- patterns:
-  - pattern: make_tainted($X)
-  - focus-metavariable: $X
+  - patterns:
+      - pattern: make_tainted($X)
+      - focus-metavariable: $X
+    by-side-effect: true
 ```
 
-**Deprecated workaround**: The previously recommended workaround was to write a source specification such as the one specified below. This instructs Semgrep that **every** occurrence of `$X` after `make_tainted($X)` must be considered a source.
+When this option is enabled, and the source specification matches a variable (or in general, an l-value) exactly, then Semgrep assumes that the variable (or l-value) becomes tainted by side-effect at the precise places where the source specification produces a match.
```

Comment:
```suggestion
When this option is enabled, and the source specification matches a variable (or in general, an [l-value](https://en.wikipedia.org/wiki/Value_(computer_science)#lrvalue)) exactly, then Semgrep assumes that the variable (or l-value) becomes tainted by side-effect at the precise places where the source specification produces a match.
```

Reply:
An `l-value` essentially means "an expression that denotes an object in memory", a memory location, something that you can use in the LHS of an assignment. Maybe we should put the definition in the text too.

Reply:
I've added this definition to https://github.com/semgrep/semgrep-docs/pull/1155